lint: Space after "!" not enforced

Looking at a PR to server code:

![image](https://cloud.githubusercontent.com/assets/128755/8524068/c800d810-23c6-11e5-8ac6-e16f340d5b96.png)

We should have eslint or jscs catch this for us. @pdehaan @TDA ideas?

Meh, will require more investigation... It does indeed look like the correct ESLint rule, but it also seemingly wants me to add spaces around `-` and `++`:

```
$ grunt eslint
Running "eslint:grunt" (eslint) task

grunttasks/connect-fonts.js
  36:75  error  Unary operator "-" must be followed by whitespace  space-unary-ops

grunttasks/l10n-supported-locales.js
  23:8   error  Space is required before unary expressions "++"    space-unary-ops
  28:8   error  Space is required before unary expressions "++"    space-unary-ops
  41:10  error  Unary operator "!" must be followed by whitespace  space-unary-ops
  46:10  error  Unary operator "!" must be followed by whitespace  space-unary-ops

✖ 5 problems (5 errors, 0 warnings)

Warning: Task "eslint:grunt" failed. Use --force to continue.

Aborted due to warnings.
```

So instead of `count++;` it now wants `count ++;`.

Comically, changing those to pass ESLint's quirkly rule logic, causes them to fail JSCS linting. Also, after fixing those 5 ESLint errors, the next subtask gives us 95 more.

--- a/.jscsrc
+++ b/.jscsrc
@@ -11,7 +11,7 @@
     "requireLineFeedAtFileEnd": true,
     "requireSpaceAfterBinaryOperators": ["=", ",", "+", "-", "/", "*", "==", "===", "!=", "!=="],
     "requireSpaceAfterKeywords": ["if", "else", "for", "function", "while", "do", "switch", "return"],
-    "requireSpaceAfterPrefixUnaryOperators": ["~"],
+    "requireSpaceAfterPrefixUnaryOperators": ["~", "!"],
     "requireSpaceBeforeBinaryOperators": ["+", "-", "/", "*", "=", "==", "===", "!=", "!=="],
     "requireSpacesInConditionalExpression": true,
     "validateIndentation": 2,

--- a/app/scripts/lib/mailcheck.js
+++ b/app/scripts/lib/mailcheck.js
@@ -37,7 +37,7 @@ function (Tooltip, Url) {
    */
   return function checkMailInput(target, metrics, translator, queryParams) {
     var element = $(target);
-    if (!element.length) {
+    if (! element.length) {
       return;
     }
     var enabled = false;

--- a/app/scripts/lib/storage-metrics.js
+++ b/app/scripts/lib/storage-metrics.js
@@ -25,7 +25,7 @@ define([
     _send: function (data) {
       var metrics = storage.get('metrics_all');
 
-      if (!Array.isArray(metrics)) {
+      if (! Array.isArray(metrics)) {
         metrics = [];
       }
 

--- a/app/scripts/models/auth_brokers/web-channel.js
+++ b/app/scripts/models/auth_brokers/web-channel.js
@@ -130,7 +130,7 @@ function (_, OAuthAuthenticationBroker, ChannelMixin, p,
 
     hasPendingOAuthFlow: function () {
       this.session.reload();
-      return !!(this.session.oauth);
+      return !! (this.session.oauth);
     },
 
     afterSignUpConfirmationPoll: function (account) {

--- a/app/scripts/models/reliers/fx-desktop.js
+++ b/app/scripts/models/reliers/fx-desktop.js
@@ -77,7 +77,7 @@ define([
      * Check if the relier wants to force the customize sync checkbox on
      */
     isCustomizeSyncChecked: function () {
-      return !!(this.isSync() && this.get('customizeSync'));
+      return !! (this.isSync() && this.get('customizeSync'));
     }
   });
 

--- a/app/scripts/views/base.js
+++ b/app/scripts/views/base.js
@@ -367,7 +367,7 @@ function (Cocktail, _, Backbone, Raven, $, p, AuthErrors,
     },
 
     trackSubview: function (view) {
-      if (!_.contains(this.subviews, view)) {
+      if (! _.contains(this.subviews, view)) {
         this.subviews.push(view);
         view.on('destroyed', _.bind(this.untrackSubview, this, view));
       }
@@ -416,7 +416,7 @@ function (Cocktail, _, Backbone, Raven, $, p, AuthErrors,
      * Return true if the success message is visible
      */
     isSuccessVisible: function () {
-      return !!this._isSuccessVisible;
+      return !! this._isSuccessVisible;
     },
 
     /**
@@ -551,7 +551,7 @@ function (Cocktail, _, Backbone, Raven, $, p, AuthErrors,
     },
 
     isErrorVisible: function () {
-      return !!this._isErrorVisible;
+      return !! this._isErrorVisible;
     },
 
     /**

--- a/app/scripts/views/form.js
+++ b/app/scripts/views/form.js
@@ -130,7 +130,7 @@ function (_, $, p, Validate, AuthErrors, BaseView, Tooltip,
 
     _isFormEnabled: true,
     isFormEnabled: function () {
-      return !!this._isFormEnabled;
+      return !! this._isFormEnabled;
     },
 
     /**

--- a/app/scripts/views/marketing_snippet_ios.js
+++ b/app/scripts/views/marketing_snippet_ios.js
@@ -67,7 +67,7 @@ define([
         showSignUpMarketing: shouldShowMarketing,
         isIos: isIos,
         isAndroid: isAndroid,
-        isOther: (!isIos && !isAndroid),
+        isOther: (! isIos && ! isAndroid),
         playStoreImage: playStoreImage
       };
     },
@@ -96,7 +96,7 @@ define([
     _isHighRes: function () {
       var win = this.window;
 
-      return !!(win.matchMedia && win.matchMedia('(-webkit-min-device-pixel-ratio: 1.5), (min-resolution: 1.5dppx), (min-resolution: 144dpi)'));
+      return !! (win.matchMedia && win.matchMedia('(-webkit-min-device-pixel-ratio: 1.5), (min-resolution: 1.5dppx), (min-resolution: 144dpi)'));
     }
   });
 

--- a/app/scripts/views/progress_indicator.js
+++ b/app/scripts/views/progress_indicator.js
@@ -141,7 +141,7 @@ define([
      * Is the progress indicator either visible or waiting to become visible?
      */
     isVisible: function () {
-      return !!this._count;
+      return !! this._count;
     }
   });
 

--- a/app/scripts/views/settings.js
+++ b/app/scripts/views/settings.js
@@ -33,7 +33,7 @@ function (Cocktail, Session, FormView, BaseView, AvatarMixin,
 
       return {
         email: email,
-        showSignOut: !account.isFromSync(),
+        showSignOut: ! account.isFromSync(),
         communicationPrefsVisible: this._areCommunicationPrefsVisible()
       };
     },

--- a/app/scripts/views/settings/communication_preferences.js
+++ b/app/scripts/views/settings/communication_preferences.js
@@ -73,7 +73,7 @@ function (Cocktail, Xss, Constants, MarketingEmailErrors, BaseView, FormView,
     submit: function () {
       var self = this;
       var emailPrefs = self.getMarketingEmailPrefs();
-      return self.setOptInStatus(NEWSLETTER_ID, !emailPrefs.isOptedIn(NEWSLETTER_ID));
+      return self.setOptInStatus(NEWSLETTER_ID, ! emailPrefs.isOptedIn(NEWSLETTER_ID));
     },
 
     setOptInStatus: function (newsletterId, isOptedIn) {

--- a/app/tests/mocks/dom-event.js
+++ b/app/tests/mocks/dom-event.js
@@ -18,7 +18,7 @@ define([
     },
 
     isDefaultPrevented: function () {
-      return !!this._defaultPrevented;
+      return !! this._defaultPrevented;
     },
 
     stopPropagation: function () {
@@ -26,7 +26,7 @@ define([
     },
 
     isPropagationStopped: function () {
-      return !!this._propagationStopped;
+      return !! this._propagationStopped;
     }
   };
 

--- a/app/tests/mocks/window.js
+++ b/app/tests/mocks/window.js
@@ -145,7 +145,7 @@ function (Backbone, sinon, _, NullStorage) {
     },
 
     isTimeoutSet: function () {
-      return !!this._isTimeoutSet;
+      return !! this._isTimeoutSet;
     },
 
     clearTimeout: function (/*timeout*/) {

--- a/app/tests/spec/models/reliers/oauth.js
+++ b/app/tests/spec/models/reliers/oauth.js
@@ -301,7 +301,7 @@ define([
           });
       });
 
-      it('!isTrusted when `trusted` is false', function () {
+      it('! isTrusted when `trusted` is false', function () {
         windowMock.location.search = TestHelpers.toSearchString({
           client_id: CLIENT_ID,
           scope: SCOPE

--- a/app/tests/spec/views/oauth_sign_up.js
+++ b/app/tests/spec/views/oauth_sign_up.js
@@ -37,7 +37,7 @@ function (chai, $, sinon, View, p, Session, FxaClient, Metrics, OAuthClient,
     context.$('[type=email]').val(email);
     context.$('[type=password]').val(password);
 
-    if (!opts.ignoreYear) {
+    if (! opts.ignoreYear) {
       $('#fxa-age-year').val(year);
     }
 

--- a/grunttasks/l10n-supported-locales.js
+++ b/grunttasks/l10n-supported-locales.js
@@ -38,12 +38,12 @@ module.exports = function (grunt) {
     var missing = [];
 
     clientKeys.forEach(function (key) {
-      if (!clientJson[key]) {
+      if (! clientJson[key]) {
         missing.push('client.json: ' + key);
       }
     });
     serverKeys.forEach(function (key) {
-      if (!messagesJson[key]) {
+      if (! messagesJson[key]) {
         missing.push('messages.json: ' + key);
       }
     });

--- a/server/lib/configuration.js
+++ b/server/lib/configuration.js
@@ -357,7 +357,7 @@ if (! process.env.NODE_ENV) {
   process.env.NODE_ENV = conf.get('env');
 }
 
-if (!conf.has('public_url')) {
+if (! conf.has('public_url')) {
   conf.set('public_url', 'https://' + conf.get('issuer'));
 }
 

--- a/server/lib/i18n.js
+++ b/server/lib/i18n.js
@@ -92,7 +92,7 @@ module.exports = function (config) {
 
   // Export the bestLanguage() function, but using defaults from the config.
   abideObj.bestLanguage = function (accepted, supported) {
-    if (!supported) {
+    if (! supported) {
       supported = config.supportedLanguages;
     }
     return abide.bestLanguage(accepted, supported, config.defaultLang);
@@ -113,7 +113,7 @@ module.exports = function (config) {
     }
     var callWasSynchronous = false;
     abideObj(fakeReq, fakeResp, function () { callWasSynchronous = true; });
-    if (!callWasSynchronous) {
+    if (! callWasSynchronous) {
       throw new Error('uh-oh, the call to i18n-abide was not synchronous!');
     }
     var l10n = {};

--- a/server/lib/routes/get-config.js
+++ b/server/lib/routes/get-config.js
@@ -40,7 +40,7 @@ module.exports = function () {
       allowedParentOrigins: allowedParentOrigins,
       authServerUrl: authServerUrl,
       env: env,
-      cookiesEnabled: !!req.cookies['__cookie_check'],
+      cookiesEnabled: !! req.cookies['__cookie_check'],
       marketingEmailServerUrl: marketingEmailApiServerUrl,
       marketingEmailPreferencesUrl: marketingEmailPreferencesUrl,
       oAuthClientId: clientId,

--- a/tests/intern.js
+++ b/tests/intern.js
@@ -19,7 +19,7 @@ function (intern, topic, firefoxProfile) {
   var fxaOauthApp = args.fxaOauthApp || 'http://127.0.0.1:8080/';
   var fxaUntrustedOauthApp = args.fxaUntrustedOauthApp || 'http://127.0.0.1:10139/';
   var fxaIframeOauthApp = args.fxaIframeOauthApp || 'http://127.0.0.1:8080/iframe';
-  var fxaProduction = !!args.fxaProduction;
+  var fxaProduction = !! args.fxaProduction;
   var fxaToken = args.fxaToken || 'http://';
   var asyncTimeout = parseInt(args.asyncTimeout || 5000, 10);
 

--- a/tests/lib/restmail.js
+++ b/tests/lib/restmail.js
@@ -7,7 +7,7 @@ define([
   'intern/node_modules/dojo/Promise'
 ], function (request, Promise) {
   function waitForEmail(uri, number) {
-    if (!number) {
+    if (! number) {
       number = 1;
     }
     console.log('Waiting for email at:', uri);

--- a/tests/server/lang.js
+++ b/tests/server/lang.js
@@ -33,7 +33,7 @@ define([
     suite['#https get ' + httpsUrl + '/ -> ' + lang] = function () {
       var dfd = this.async(intern.config.asyncTimeout);
       request(httpsUrl + '/', options, dfd.callback(function (err, res) {
-        assert.ok(!err);
+        assert.ok(! err);
         assert.equal(res.statusCode, 200);
         var re = new RegExp(util.format('lang="%s"', lang));
         assert.ok(res.body.match(re), 'html has correct lang attribute');
@@ -43,7 +43,7 @@ define([
     suite['#https get ' + httpsUrl + '/i18n/client.json -> ' + lang] = function () {
       var dfd = this.async(intern.config.asyncTimeout);
       request(httpsUrl + '/i18n/client.json', options, dfd.callback(function (err, res) {
-        assert.ok(!err);
+        assert.ok(! err);
         assert.equal(res.statusCode, 200);
         if (intern.config.fxaProduction) {
           // using the empty string '' as the key below is intentional

--- a/tests/server/routes.js
+++ b/tests/server/routes.js
@@ -73,7 +73,7 @@ define([
     routes['/503.html'] = { statusCode: 200 };
   }
 
-  if (!intern.config.fxaProduction) {
+  if (! intern.config.fxaProduction) {
     routes['/tests/index.html'] = { statusCode: 200 };
     routes['/tests/index.html?coverage'] = { statusCode: 200 };
     routes['/boom'] = { statusCode: 500 };

